engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/types/TypesInformationExtractor.scala

```diff
@@ -62,40 +60,84 @@ object TypesInformationExtractor {
     case TypedTaggedValue(underlying, _) =>
       clazzRefsFromTypedClass(underlying.objType)
     case Unknown =>
-      Set()
+      Set.empty
   }
 
   private def clazzRefsFromTypedClass(typedClass: TypedClass): Set[ClazzRef]
     = typedClass.params.flatMap(clazzRefsFromTypingResult).toSet + ClazzRef(typedClass.klass)
 
-  private def clazzAndItsChildrenDefinition(clazzRef: ClazzRef, currentSet: Set[ClazzRef])
-                                   (implicit settings: ClassExtractionSettings): Set[ClazzDefinition] = {
-    val clazz = clazzRef.clazz
-    val newSet = currentSet + clazzRef
+  private def clazzAndItsChildrenDefinition(clazzRef: ClazzRef)
+                                           (collectedSoFar: mutable.Set[ClazzRef], path: DiscoveryPath)
+                                           (implicit settings: ClassExtractionSettings): Set[ClazzDefinition] = {
+    // we not support arrays for now - will looks ugly in UI, so we need to hide it
+    if (clazzRef.clazz.isArray || collectedSoFar.contains(clazzRef)) {
```

Comment:
why we handle arrays here and not in isHidden?

Reply:
I would feel better if this "if" were directly in flatMap in line 47. We would use mutable stuff only in that method and not here

Reply:
1. I will try to extract it to predicates
2. Good point, let's keep this mutable stuff in separate place. Will change it.

Reply:
extracted + extracted mutable stuff to one method